### DISCUSSION TITLE
Can't find proper way to deal with events from external processes

### DISCUSSION THREAD

**Discussion by l1onsun** [fastapi/fastapi#8945] [Questions] [answered]
### Description
I need to handle events occurring in worker process. My worker works continuously and sometimes updates database.  When this happens, i want to notify clients that data has been updated. 
And i can't find proper way to notify my fastapi-app from worker.
### Simplified Example

```Python
# main.py
from fastapi import FastAPI, Websockets
app = FastAPI()

class ConnectionManager():...
connection_manager = ConnectionManager()

def on_worker_event(updated_data): # the question is how/where to register this callback
    connection_manager.notify_clients(updated_data)

@app.websocket("/ws")
async def websocket_endpoint(websocket: Websocket):
    await websocket.accept()
    connection_manager.add_client(websocket)
```
```Python
# worker.py
while True:
    # do some work
    updated_data = updated_data()
    # somehow notify `main` that updated_data ready
```
### My current solution
```Python
# in main.py
@app.post('/event')
def on_worker_event(updated_data = Body(...))
    connection_manager.notify_clients(updated_data)

# in worker.py
while True:
    ...
    requests.post("http://127.0.0.1:8000/event", data=updated_data) 
```
There is big disadvantage - i can use only one process, otherwise not all clients will be notified.
To solve this i can start several Uvicorn processes on different ports and use for example Nginx for load-balancing. But i don't sure this is right pattern.

Or may be i need use some event queue, but i can't find a way to listen it from `main.py` without blocking the server.

**Comment by l1onsun** [comment]
found related issue https://github.com/tiangolo/fastapi/issues/825

Would this be the right pattern?
```Python
@app.on_event("startup")
async def startup():
    loop = asyncio.get_running_loop()
    # somehow, using loop, start async listening of my event queue
    # (i currently don't know how, but i think i'll find way)
```

**Comment by l1onsun** [comment]
I thought i found solution, but now again i'm not sure if it is good designed.

 I created intermediate layer (called say `notify_router`)  between main servers and workers. My `main` processes register in `notify_router` on "startup" by creating websocket connection. Now when event happens in `worker` it sends it to `notify_router` and `notify_router` resend event to desired `main` processes.

The problem is -> to create websocket connection from my main.py i use:
```Python
@app.on_event("startup")
async def register():
    asyncio.get_running_loop().create_task(connect_to_notify_router())
```
And I'm not sure if this is correct

**Comment by pro4jhu** [comment]
hi, man! Did you solved? i am confusing the same thing.
i think uvicorn start app by multi processes. every process has its own data scope. the cannot share data because uvicorn spawn process without passing other customize data. event loop seems one thread has one event loop. i tested use @app.on_event("startup") ,but when run uvicorn mutli workers it run multi times. i think solution is use other data tool, like redis, kafka. then you websocket loop to listen data change and push to client. 
 i have checked django-channels , it has layer, layer can group send. it use redis to layer backend. i do not know how to implement it. django-channels has in-Memory  layer, but it only work in one process. check this [https://channels.readthedocs.io/en/latest/topics/channel_layers.html#in-memory-channel-layer](url), i think in production, django-channels use redis to save message, and websocket loop to listen it and push it to client.
i have checked this, it may say how it works [https://channels.readthedocs.io/en/latest/topics/channel_layers.html#what-to-send-over-the-channel-layer](url)

**Comment by YuriiMotov** [comment, marked as answer]
Proposed approach will only work (?) if you have one instance of your app (one uvicorn worker).
In general you need to use any of distributed systems like Redis, RabbitMQ, Kafka (or others) to implement the publish\subscribe mechanism
